genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -0,0 +1,3394 @@
+        face = self.polytope_faces[i_b, i_f]
+        face_v1 = self.polytope_verts[i_b, face.verts_idx[0]].mink
+        face_v2 = self.polytope_verts[i_b, face.verts_idx[1]].mink
+        face_v3 = self.polytope_verts[i_b, face.verts_idx[2]].mink
+        face_normal = face.normal
+
+        _lambda = self.func_triangle_affine_coords(
+            face_normal,
+            face_v1,
+            face_v2,
+            face_v3,
+        )
+
+        # Point on geom 1
+        v1 = self.polytope_verts[i_b, face.verts_idx[0]].obj1
+        v2 = self.polytope_verts[i_b, face.verts_idx[1]].obj1
+        v3 = self.polytope_verts[i_b, face.verts_idx[2]].obj1
+        witness1 = v1 * _lambda[0] + v2 * _lambda[1] + v3 * _lambda[2]
+
+        # Point on geom 2
+        v1 = self.polytope_verts[i_b, face.verts_idx[0]].obj2
+        v2 = self.polytope_verts[i_b, face.verts_idx[1]].obj2
+        v3 = self.polytope_verts[i_b, face.verts_idx[2]].obj2
+        witness2 = v1 * _lambda[0] + v2 * _lambda[1] + v3 * _lambda[2]
+
+        self.witness[i_b, 0].point_obj1 = witness1
+        self.witness[i_b, 0].point_obj2 = witness2
+
+        # Safe measure for mujoco consistency
+        normal = witness1 - witness2
+        normal_length = ti.sqrt(normal.dot(normal))
+        if normal_length < self.FLOAT_MIN:
+            normal = gs.ti_vec3(1.0, 0.0, 0.0)  # Default normal if zero length
+        else:
+            normal = normal * (1.0 / normal_length)  # Normalize the normal vector
+        self.normal[i_b] = normal
+
+    @ti.func
+    def func_epa_horizon(self, i_b, nearest_i_f):
+        """
+        Compute the horizon, which represents the area of the polytope
+        that is visible from the vertex w, and thus should be deleted
+        for the expansion of the polytope.
+        """
+        w = self.polytope[i_b].horizon_w
+
+        # Initialize the stack by inserting the nearest face
+        self.polytope_horizon_stack[i_b, 0].face_idx = nearest_i_f
+        self.polytope_horizon_stack[i_b, 0].edge_idx = 0
+        top = 1
+        is_first = True
+
+        flag = 0
+        while top > 0:
+
+            # Pop the top face from the stack
+            i_f = self.polytope_horizon_stack[i_b, top - 1].face_idx
+            i_e = self.polytope_horizon_stack[i_b, top - 1].edge_idx
+            top -= 1
+
+            # If the face is already deleted, skip it
+            is_deleted = self.polytope_faces[i_b, i_f].map_idx == -2
+            if (not is_first) and (is_deleted):
+                continue
+
+            face = self.polytope_faces[i_b, i_f]
+
+            # Check visibility of the face.
+            # Two requirements for the face to be visible:
+            # 1. The face normal should point towards the vertex w
+            # 2. The vertex w should be on the other side of the face
+            #    to the origin
+            is_visible = face.normal.dot(w) - face.dist2 > self.FLOAT_MIN
+
+            # The first face is always visible.
```

Comment:
`# The first face is always CONSIDERED visible.` because it is was, `is_first` would be irrelevant in the first place.